### PR TITLE
WIP: Support for weak dependencies

### PR DESCRIPTION
**Pull request by KristofferC**
See https://github.com/JuliaLang/julia/pull/47040

### CODE REVIEW

File: docs/src/weakdeps.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# Weak dependencies
+
+It is sometimes desirable to be able to extend some functionality of a package without having to
+unconditionally take on the cost (in terms of e.g. load time) of adding an extra dependency.
+A *weak* dependency is a package that is only available to load if some other package in the
+current environment have that package as a normal (or strong) dependency.
```
Comment (Seelengrab) [inline docs/src/weakdeps.md:6]:
```suggestion
current environment has that package as a normal (or strong) dependency. 
```

File: docs/src/weakdeps.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
+# Weak dependencies
+
+It is sometimes desirable to be able to extend some functionality of a package without having to
+unconditionally take on the cost (in terms of e.g. load time) of adding an extra dependency.
+A *weak* dependency is a package that is only available to load if some other package in the
+current environment have that package as a normal (or strong) dependency. 
+
+Weak dependencies are listed in a `Project.toml` file under the `[weakdeps]` section which can be compared to a
+(strong) dependency which is under the `[deps]` section.
+Compatibility on weak dependencies are specified like a normal dependeny unde the ´[compat]` section.
```
Comment (Seelengrab) [inline docs/src/weakdeps.md:10]:
```suggestion
Compatibility on weak dependencies are specified like a normal dependency under the `[compat]` section.
```

File: docs/src/weakdeps.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+`src/Plotting.jl`:
+```julia
+module Plotting
+
+using Colors
+
+if Base.@hasdep Contour
+    using Contour
+    function plot(c::Contour.ContourCollection
```
Comment (Seelengrab) [inline docs/src/weakdeps.md:44]:
```suggestion
    function plot(c::Contour.ContourCollection)
```

File: docs/src/weakdeps.md (hunk 2 above)
Comment (KristofferC) [reply docs/src/weakdeps.md:10]: Thanks, the current docs are still very WIP.

File: docs/src/weakdeps.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    end
+end
+
+end # module
+```
+
+## Compatibility with older Julia versions.
+
+It is possible to have a dependency be a weak version in Julia versions that support it and be a strong dependency in earlier
+Julia versions. This is done by having the dependency as *both* a strong and weak dependency. Older Julia version will ignore
```
Comment (Seelengrab) [inline docs/src/weakdeps.md:55]:
```suggestion
Julia versions. This is done by having the dependency as *both* a strong and weak dependency. Older Julia versions will ignore
```

File: docs/src/weakdeps.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+the specificaton of the depenency as weak while new Julia versions will tag it as a weak dependency.
```
Comment (Seelengrab) [inline docs/src/weakdeps.md:56]:
```suggestion
the specification of the dependency as weak while new Julia versions will tag it as a weak dependency.
```

File: src/API.jl [hunk 6]
```diff
@@ -301,9 +301,39 @@ function rm(ctx::Context, pkgs::Vector{PackageSpec}; mode=PKGMODE_PROJECT, all_p
...
+    dep_list = String[]
+    new_uuids = keys(ctx.env.manifest)
+    old_uuids = keys(ctx.env.original_manifest)
+    for (uuid, pkgentry) in ctx.env.manifest.deps
+        (isnothing(pkgentry.weakdeps) || isempty(pkgentry.weakdeps)) && continue
+        weak_uuids = values(pkgentry.weakdeps)
+        old_num_promoted = count(in(old_uuids), weak_uuids)
+        new_num_promoted = count(in(new_uuids), weak_uuids)
+
+        # @show pkgentry pkgentry.weakdeps old_uuids new_uuids new_num_promoted old_num_promoted
```
Comment (Seelengrab) [inline src/API.jl:327]:
```suggestion
```

**Comment by IanButterworth** [comment]
Closing because package extensions won #3264

**Pull request closed without merging**